### PR TITLE
Implement RotatE

### PR DESCRIPTION
**Pull request by huonw**
This implements the core of the RotatE algorithm from "RotatE: Knowledge Graph Embedding by Relational Rotation in Complex Space" Zhiqing Sun, Zhi-Hong Deng, Jian-Yun Nie, Jian Tang https://arxiv.org/abs/1902.10197

That core is the scoring function and model, as well as some basic tests.

I intend to generalise this model to share more code with `ComplEx` and `DistMult` (#1549). Documentation can happen as part of that effort.

#1550 tracks adding a demo notebook, which can happen in parallel with #1549.

See: #1521

### CODE REVIEW

File: stellargraph/layer/knowledge_graph.py [hunk 1]
```diff
@@ -529,6 +529,164 @@ def in_out_tensors(self):
...
+        r_re = tf.math.cos(r_phase)
+        r_im = tf.math.sin(r_phase)
+
+        o_re = self._node_embeddings_real(o_iloc)
+        o_im = self._node_embeddings_imag(o_iloc)
+
+        return self._scoring([s_re, s_im, r_re, r_im, o_re, o_im])
+
+    def in_out_tensors(self):
+        s_iloc = Input(shape=1)
```
Comment (kieranricardo) [inline stellargraph/layer/knowledge_graph.py:680]: Is this extra dimension `shape=1` needed here? From what I can tell this dimension isn't being used, could this instead be:

```suggestion
        s_iloc = Input(batch_shape=(None,))
``

File: stellargraph/layer/knowledge_graph.py [hunk 2]
```diff
@@ -529,6 +529,164 @@ def in_out_tensors(self):
...
+    def get_config(self):
+        return {**super().get_config(), "margin": self._margin}
+
+    def call(self, inputs):
+        s_re, s_im, r_re, r_im, o_re, o_im = inputs
+        # expansion of s◦r - t
+        re = s_re * r_re - s_im * r_im - o_re
+        im = s_re * r_im + s_im * r_re - o_im
+        # norm the vector: -|| ... ||_2^2
+        return self._margin - tf.reduce_sum(re * re + im * im, axis=2)
```
[the remark this answers is not in the document]
Comment (huonw) [reply stellargraph/layer/knowledge_graph.py:546]: Hm, good catch. I don't know where I got norm-squared from, but I did explicitly switch to it for some reason (https://github.com/stellargraph/stellargraph/pull/1522/commits/8922d19de94b98b5f4da68a64994445072b75805)... but clearly that didn't make sense at all.

Despite what the paper says, it turns out their code seems to be using L2 norm: 
- https://github.com/DeepGraphLearning/KnowledgeGraphEmbedding/blob/2442e8b9bb1a40e44edabcb3cc4f145ac24b1c7b/codes/model.py#L225
- https://pytorch.org/docs/master/generated/torch.norm.html

But it definitely seems like their paper is using L1 norm, e.g. the explicit expansion in Appendix E on page 13:

image

I might split the difference by adding a `norm_order` parameter and using `tf.norm(..., ord=norm_order)`, defaulting to ... 1 or 2??

[the remark this answers is not in the document]
Comment (kieranricardo) [reply stellargraph/layer/knowledge_graph.py:546]: huh... I think that definition of (8) is at odds with their initial definition:

image

According to that eq 8 should be 

```python
sum(abs(re(h * r) - re(t)) + abs(im(h * r) - im(t)))
```

Like the norm applied to each complex number should be L1 not L2!

File: stellargraph/layer/knowledge_graph.py (hunk 1 above)
Comment (huonw) [reply stellargraph/layer/knowledge_graph.py:680]: That's potentially true, although it makes this differ to `ComplEx` and `DistMult`. As such, I'd prefer to keep it as is, and change them all separately?

File: stellargraph/layer/knowledge_graph.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (kieranricardo) [reply stellargraph/layer/knowledge_graph.py:546]: yeah I think that sounds reasonable, maybe a default of 2 so its easy to replicate the paper's results - what do you think?

File: stellargraph/layer/knowledge_graph.py (hunk 1 above)
Comment (kieranricardo) [reply stellargraph/layer/knowledge_graph.py:695]: yeah fair point, just something to flag if you're gonna go ahead with that refactor of all 3 but its pretty minor :)

File: stellargraph/layer/knowledge_graph.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (huonw) [reply stellargraph/layer/knowledge_graph.py]: Sounds good. I went with that.

File: stellargraph/layer/knowledge_graph.py (hunk 1 above)
Comment (huonw) [reply stellargraph/layer/knowledge_graph.py:695]: I made a note on #1549.
